Add design document for Mobile runtimeconfig.json host configuration (#50744)
dotnet/runtime: docs/design/mono/mobile-runtimeconfig-json.md

# Mobile runtimeconfig.json host configuration

## Motivation

> `runtimeconfig.json` is a file used by .NET 5+ applications to pass arbitrary configuration parameters via the application host to the .NET runtime.  On mobile and WebAssembly, we would like to support passing configuration properties from the Mono embedder to the runtime.

Reviewer: Do we plan to adopt this on wasm? I thought this was intended to be mobile-specific for now? cc: @marek-safar

Reviewer 2: I think when I was drafting this part "mobile and WebAssembly" was just shorthand for "workloads where Mono is the runtime"

Author: Quote from Marek "It should be supported by all Mono workloads, in other words passing runtimeconfig.json settings have to always work"

> To minimize the impact of the app startup time, the design should preserve these contraints:

Reviewer: ```suggestion
To minimize the impact on the app startup time, the design constraints are as follows:
```

> 1. Don’t parse JSON file at runtime at application startup

Reviewer: ```suggestion
1. Don’t parse JSON at runtime during application startup
```

> 2. Don’t need to support full set of well-known properties, just the ones that make sense on mobile

Reviewer: I think you can get rid of this, we support everything

Author: Removed item 2

3. Don’t need to support shared framework configuration parameters
4. Don't need to support the desktop host environment variables, startup hooks, or other configuration mechanisms that do not make sense on mobile.
5. This is separate from the key/value set passed to `monovm_initialize`/`coreclr_initialize`  (ie we will not pass the additional properties via `monovm_initialize` - just using this new mechanism). Things like the TPA list, additional probing paths, PINVOKE_OVERRIDE, etc will be addressed by an evolution of `monovm_initialize` - https://github.com/dotnet/runtime/issues/48416

## Design Overview

We break up runtimeconfig.json loading into two parts:

> 1. A new MSBuild task called `RuntimeConfigParser` will run after the `runtimeconfig.json` is created by the dotnet build process. The task will extract properties keys and values into a binary blob format. The resulting `runtimeconfig.blob` file will be bundled with the application.

Reviewer: ```suggestion
1. A new MSBuild task called `RuntimeConfigParser` will run after the `runtimeconfig.json` is created by the `dotnet build` process. The task will convert the properties and their values into a binary blob format. The resulting `runtimeconfig.blob` file will be bundled with the application.
```

Also, did we actually end up using ".blob" as the extension? That's pretty funny, though not a problem.

Author: We use runtimeconfig.bin actually. Let me change this.

> 2. The runtime will expose a new API entrypoint `monovm_runtimeconfig_initialize` that gets either a path to pass to `mono_file_map_open` or a pointer to the blob in memory. Then, the runtime will read the binary data and populate the managed AppContext with the properties.

Reviewer: ```suggestion
2. The runtime will expose a new API entrypoint `monovm_runtimeconfig_initialize` that gets either a path to pass to `mono_file_map_open` or a pointer to the blob in memory. When called, the runtime will read the binary data and populate the managed AppContext with the properties.
```

> We will only take the `runtimeOptions→configProperties` json key. Its content is a json dictionary with string keys and string/bool/numeric values.  We convert the values to strings when we store them in the binary runtimeconfig.blob.

Reviewer: ```suggestion
We will only use the `runtimeOptions→configProperties` json key. Its content is a JSON dictionary with string keys and string/bool/numeric values.  We convert the values to strings when we store them in the binary runtimeconfig.blob, which is the same way they are treated by the default host.
```

> The runtime will assume that the properties passed via `monovm_initialize` and `monovm_runtimeconfig_initialize` will be different. To ensure this, the MSBuild task that we will provide will be given a list of property names that the embedder promises it will pass to `monovm_initialize`. The MSBuild task will check that `runtimeconfig.json` does not set any of those same properties. If there is a duplicate, error out.

Reviewer: ```suggestion
The runtime assumes that the properties passed via `monovm_initialize` and `monovm_runtimeconfig_initialize` will be different. To ensure this, the provided MSBuild task will be passed a list of property names that the embedder promises it will pass to `monovm_initialize`. The MSBuild task will check that `runtimeconfig.json` does not set any of those same properties. If there is a duplicate, error out.
```

> We take everything and pass all the properties to the managed AppContext. For the well-known standard properties the mono runtime will read and use the ones it cares about.

Reviewer: ```suggestion
All properties set in either the `runtimeconfig.json` or set via `monovm_initialize` will be propagated to the managed `AppContext`. Mono will also check for properties it supports in the runtime itself and make use of them as appropriate.
```

## Design Details

### Encoded file generation

>  The runtime pack will provide an MSBuild task called  to generate the encoded file. The generator checks for duplicate property keys (by comparing the keys in the json file with an input list of properties that the embedder promises to pass to `monovm_initialize`).

Author: Good catch!

Sample input
```
{
    "runtimeOptions": {
        "configProperties": {
            "key1": "value1",
            "key2": "value2"
        }
    }
}
```

Sample output (as hexdump -C bytes)
```
00000000  02 04 6b 65 79 31 06 76  61 6c 75 65 31 04 6b 65  |..key1.value1.ke|
00000010  79 32 06 76 61 6c 75 65  32                       |y2.value2|
00000019
```

### New embedding API entrypoint

We want at least 2 ways to pass data in: either ask the runtime to open the file, or give it a pointer to the data in memory. Also, we need some way to cleanup.

```
struct MonovmRuntimeConfigArguments {
  uint32_t kind; // 0 = Path of runtimeconfig.bin file, 1 = pointer to the blob data, >= 2 reserved
  union {
    struct {
      const char *path; // null terminated absolute path
    } name;
    struct {
      const char *data;
      uint32_t data_len;
    } data;
  } runtimeconfig;
};

typedef void (*MonovmRuntimeConfigArgumentsCleanup)(MonovmRuntimeConfigArguments *args, void* user_data);

MONO_API void
monovm_runtimeconfig_initialize (MonovmRuntimeConfigArguments *args, MonovmRuntimeConfigArgumentsCleanup cleanup_fn, void* user_data);
```

> This declaration should live in the unstable header. `monovm_runtimeconfig_initialize` should be called after `monovm_initialize` but before starting the runtime.

Reviewer: I thought we ended up going with the other way around - `monovm_runtimeconfig_initialize` before `monovm_initialize` before starting the runtime? Check to make sure this is consistent with the code, but as long as it is we should be fine.

Author: It is before. Let me fix this

#### Example of the usage of `monovm_runtimeconfig_initialize`

```
void
cleanup_runtime_config (MonovmRuntimeConfigArguments *args, void *user_data)
{
    free (args);
    free (user_data); // This may not be needed, depending on if there is anything needs to be freed.
}

int
mono_droid_runtime_init (const char* executable, int managed_argc, char* managed_argv[])
{
  
  ......

  MonovmRuntimeConfigArguments *arg = (MonovmRuntimeConfigArguments *)malloc (sizeof (MonovmRuntimeConfigArguments));
  arg->kind = 0;
  arg->runtimeconfig.name.path = "path_to_generated_binary_file";
  monovm_runtimeconfig_initialize (arg, cleanup_runtime_config, NULL);

  monovm_initialize(......);
  ......

}
```

### Register and install runtime properties

> `monovm_runtimeconfig_initialize` will register the type `MonovmRuntimeConfigArguments` variable with the runtime. If given the path of the runtimeconfig.bin file, the runtime will read the binary data from the file, otherwise, it will parse the binary data from memory. The properties will be combined with the ones registered by `monovm_initialize` and used to initialize System.AppContext.

Reviewer: ```suggestion
`monovm_runtimeconfig_initialize` will register the type `MonovmRuntimeConfigArguments` variable with the runtime. If given the path of the runtimeconfig.bin file, the runtime will read the binary data from the file, otherwise, it will parse the binary data from memory. The properties will be combined with the ones registered by `monovm_initialize` and used to initialize System.AppContext.
```

This says `runtimeconfig.bin`, and elsewhere in the document we refer to blob. We should be consistent, and either call it "blob" or "binary" within the document based on the extension. I weakly prefer ".bin" for the extension, but it doesn't matter very much.

Author: yes `.bin` is the extension that I used in the test. Will update.

## Motivation

`runtimeconfig.json` is a file used by .NET 5+ applications to pass arbitrary configuration parameters via the application host to the .NET runtime.  On mobile and WebAssembly, we would like to support passing configuration properties from the Mono embedder to the runtime.

To minimize the impact of the app startup time, the design should preserve these contraints:
1. Don’t parse JSON file at runtime at application startup
2. Don’t need to support full set of well-known properties, just the ones that make sense on mobile
3. Don’t need to support shared framework configuration parameters
4. Don't need to support the desktop host environment variables, startup hooks, or other configuration mechanisms that do not make sense on mobile.
5. This is separate from the key/value set passed to `monovm_initialize`/`coreclr_initialize`  (ie we will not pass the additional properties via `monovm_initialize` - just using this new mechanism). Things like the TPA list, additional probing paths, PINVOKE_OVERRIDE, etc will be addressed by an evolution of `monovm_initialize` - https://github.com/dotnet/runtime/issues/48416

## Design Overview

> This feature could be achived by mainly two parts:

Reviewer 2: nit: "achieved".  but even better 

```suggestion
We break up runtimeconfig.json loading into two parts:
```

1. A new MSBuild task called `RuntimeConfigParser` will run after the `runtimeconfig.json` is created by the dotnet build process. The task will extract properties keys and values into a binary blob format. The resulting `runtimeconfig.blob` file will be bundled with the application.
2. The runtime will expose a new API entrypoint `monovm_runtimeconfig_initialize` that gets either a path to pass to `mono_file_map_open` or a pointer to the blob in memory. Then, the runtime will read the binary data and populate the managed AppContext with the properties.

> We will only take the `runtimeOptions→configProperties` json key. Its content is a flat key-value string/boolean dictionary.

Reviewer 2: ```suggestion
We will only take the `runtimeOptions→configProperties` json key. Its content is a flat key-value string/string dictionary.
```

Author: Fixed. Thanks for the suggestion!

The runtime will assume that the properties passed via `monovm_initialize` and `monovm_runtimeconfig_initialize` will be different. To ensure this, the MSBuild task that we will provide will be given a list of property names that the embedder promises it will pass to `monovm_initialize`. The MSBuild task will check that `runtimeconfig.json` does not set any of those same properties. If there is a duplicate, error out.

> We take everything and pass all the properties to the managed AppContext. For the well-known standard properties the mono runtime will read and propagate the ones it cares about and ignore the rest.

Reviewer 2: ```suggestion
We take everything and pass all the properties to the managed AppContext. For the well-known standard properties the mono runtime will read and use the ones it cares about.
```

(I think "and ignore the rest" makes it sound like we will drop them from the managed AppContex,t too.  Which is not the case)

Author: Fixed. Thanks for the suggestion!

## Design Details

### Encoded file generation

> We should be able to generate the encoded file from C# as an MSBuild task, using System.Text.Json and System.Reflection.Metadata. This generator should also do duplicate checking (by comparing the keys in the json file with an input list of properties that the embedder promises to pass to `monovm_initialize`).

Reviewer 2: ```suggestion
We should be able to generate the encoded file using an MSBuild task. This generator should also do duplicate checking (by comparing the keys in the json file with an input list of properties that the embedder promises to pass to `monovm_initialize`).
```

Author: Fixed. Thanks for the suggestion!

00000010  79 32 06 76 61 6c 75 65  32                       |y2.value2|
00000019
```

### New embedding API entrypoint

We want at least 2 ways to pass data in: either ask the runtime to open the file, or give it a pointer to the data in memory. Also, we need some way to cleanup.

```
struct MonovmRuntimeConfigArguments {
  uint32_t kind; // 0 = Path of runtimeconfig.bin file, 1 = pointer to the blob data, >= 2 reserved
  union {
    struct {
      const char *path; // null terminated absolute path
    } name;
    struct {
      const char *data;
      uint32_t data_len;
    } data;
  } runtimeconfig;
};

typedef void (*MonovmRuntimeConfigArgumentsCleanup)(MonovmRuntimeConfigArguments *args, void* user_data);

MONO_API void
monovm_runtimeconfig_initialize (MonovmRuntimeConfigArguments *args, MonovmRuntimeConfigArgumentsCleanup cleanup_fn, void* user_data);
```

This declaration should live in the unstable header. `monovm_runtimeconfig_initialize` should be called after `monovm_initialize` but before starting the runtime. The cleanup function should be called in `mono_runtime_install_appctx_properties`.

### Register and install runtime properties

> `monovm_runtimeconfig_initialize` will register the type `MonovmRuntimeConfigArguments` variable as it is. Processing and installing the properties will be done inside `mono_runtime_install_appctx_properties`. If given the path of runtimeconfig.blob file, read the binary data from the file first. Otherwise, parse the binary data and combine the properties with the ones registered by `monovm_initialize`.

Reviewer 2: Add a paragraph here about `cleanup_fn`. Something like:

The `MonovmRuntimeConfigArguments*` will be stored in the runtime between `monovm_runtimeconfig_initialize` and `mono_jit_init_version`.  The embedder should not dispose of the arguments after calling `monovm_runtimeconfig_initialize`.  Instead the runtime will call the `cleanup_fn` that is passed to `monovm_runtimeconfig_initialize` at soon after it has initialized the managed property list in `System.AppContext` (which happens as part of `mono_jit_init_version`)

Author: Fixed. Thanks for the suggestion!

> `monovm_runtimeconfig_initialize` will register the type `MonovmRuntimeConfigArguments` variable as it is. Processing and installing the properties will be done inside `mono_runtime_install_appctx_properties`. If given the path of runtimeconfig.blob file, read the binary data from the file first. Otherwise, parse the binary data and combine the properties with the ones registered by `monovm_initialize`.

Reviewer 2: ```suggestion
`monovm_runtimeconfig_initialize` will register the type `MonovmRuntimeConfigArguments` variable with the runtime. If given the path of the runtimeconfig.blob file, the runtime will read the binary data from the file, otherwise, it will parse the binary data from memory. The properties will be combined with the ones registered by `monovm_initialize` and used to initialize System.AppContext.
```

Author: Fixed. Thanks for the suggestion!
